pilot/pkg/proxy/envoy/v2/ondemand.go

```diff
@@ -0,0 +1,438 @@
+								rdsReject.With(prometheus.Labels{"node": discReq.Node.Id, "err": discReq.ErrorDetail.Message}).Add(1)
+								totalXDSRejects.Add(1)
+							} else { // protocol error
+								adsLog.Warnf("ADS:RDS: ACK PROTOCOL ERROR %v %s (%v) %v", peerAddr, con.ConID, con.modelNode, discReq.String())
+								rdsReject.With(prometheus.Labels{"node": discReq.Node.Id, "err": "Protocol error"}).Add(1)
+								totalXDSRejects.Add(1)
+								continue
+							}
+							continue
+						}*/
+				}
+
+				//BAVERY_FIXME: handle unsubscribe
+				unsubscribeRoutes := discReq.GetResourceNamesUnsubscribe()
+
+				if sortedRoutes == nil {
+					sort.Strings(routes)
+					sortedRoutes = routes
+				}
+				con.Routes = sortedRoutes
+				adsLog.Debugf("ADS:RDS: REQ %s %s  routes: %d", peerAddr, con.ConID, len(con.Routes))
+				err := s.pushDeltaRoute(con, s.globalPushContext(), unsubscribeRoutes)
+				if err != nil {
+					return err
+				}
+
+			case VhdsType:
+				//BAVERY_TODO: Find a better way to handle these environment variables than parsing each time
+				if enableVHDS, _ := strconv.ParseBool(con.modelNode.Metadata["ENABLE_DYNAMIC_HOST_CONFIGURATION"]); !enableVHDS {
+					return status.Errorf(codes.Unavailable, "VHDS not enabled.")
+				}
+
+				if discReq.ErrorDetail != nil {
+					adsLog.Warnf("ADS:VHDS: ACK ERROR %v %s (%s) %v", peerAddr, con.ConID, con.modelNode.ID, discReq.String())
+					vhdsReject.With(prometheus.Labels{"node": discReq.Node.Id, "err": discReq.ErrorDetail.Message}).Add(1)
+					totalXDSRejects.Add(1)
+					continue
+				}
+
+				//BAVERY_TODO: Complete
+				//currentHosts := []string{}
+				subscribeHosts := discReq.GetResourceNamesSubscribe()
+				sort.Strings(subscribeHosts)
+				unsubscribeHosts := discReq.GetResourceNamesUnsubscribe()
+				sort.Strings(unsubscribeHosts)
+				//mergeSubscribeUnsubsribe(currentHosts, subscribeHosts, unsubscribeHosts)
+				fmt.Printf("Bavery vhds request received. \n Subscribing to: %+v \n unsubscribing from: %+v\n\n\n", subscribeHosts, unsubscribeHosts)
+
+				con.mu.RLock()
+				vHostNonceSent := con.VHostNonceSent
+				con.mu.RUnlock()
+
+				if vHostNonceSent != "" && vHostNonceSent != discReq.ResponseNonce {
+					adsLog.Debugf("ADS:RDS: Expired nonce received %s %s (%v), sent %s, received %s",
+						peerAddr, con.ConID, con.modelNode, vHostNonceSent, discReq.ResponseNonce)
+					vhdsExpiredNonce.Inc()
+					continue
+				}
+
+				//BAVERY_TODO: Add ErrorDetail
+				//BAVERY_QUESTION: Do we support namespace in the path? Probably not.
+
+				for _, subscribeVhost := range subscribeHosts {
+					routeName, host, err := splitVHost(subscribeVhost)
+					if err != nil {
+						adsLog.Errorf("Request received with invalid route config name: %+v.", err.Error())
+					} else {
+						//use wildcard namespace to select the service from any available namespace. We likely can't get the actual namespace from Envoy.
```

Comment:
What do you mean, the namespace of the client is passed as one of the metadata.

Reply:
I was referring to the fact that we can find the DNS entry to which the request is expecting to refer, but not whether the entry is expecting to be in the bookinfo, default, etc namespace.